### DISCUSSION TITLE
String to literal after parse

### DISCUSSION THREAD

**Discussion by alexey13** [sinclairzx81/typebox#716] [Q&A] [answered]
First of all thank you for the great library! Tried zod but typebox is a lot better for my use cases.

Please take a look at this example:

We have base schema

```js
const BaseComponentSchema = Type.Object({
  type: Type.String(),
  id: Type.String(),
  name: Type.String(),
});
```

And we have a method CreateComponentSchema this method should return ComponentSchema but type and id should be literals. This will work
```js
/**
 * @template {string} T
 * @template {string} B
 * @param {object} config
 * @param {T} config.id
 * @param {B} config.type
 */
const CreateComponentSchema = (config) => {
  return Type.Composite([
    BaseComponentSchema,
    Type.Object({
      id: Type.Literal(config.id),
      type: Type.Literal(config.type),
    }),
  ]);
};

const TestComponentSchema = CreateComponentSchema({
  id: 'test',
  type: 'asdasd',
});

/**@typedef {import('@sinclair/typebox').Static<typeof TestComponentSchema>} TestComponent */
// {id: 'test'...}
```

we would like to validate config first but if we are using unknown there is no way anymore to infer string value as literal, right?

```js
/**
 * @param {unknown} config
 */
const CreateComponentSchema = (config) => {
  const check = Check(BaseComponentSchema, config);
  if (!check)
    throw new Error(Errors(BaseComponentSchema, config)?.First()?.message);
  const parsedConfig = /**@type {import('@sinclair/typebox').Static<typeof BaseComponentSchema>}*/ (config);
  return Type.Composite([
    BaseComponentSchema,
    Type.Object({
      id: Type.Literal(parsedConfig.id),
      type: Type.Literal(parsedConfig.type),
    }),
  ]);
};

const TestComponentSchema = CreateComponentSchema({
  id: 'test',
  type: 'asdasd',
});

/**@typedef {import('@sinclair/typebox').Static<typeof TestComponentSchema>} TestComponent */
// {id: string...}
```
Thanks a lot in advance!

**Comment by sinclairzx81** [comment, marked as answer]
@alexey13 Hi, Try the following

[TypeScript Link Here](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgFQJ5gKYBo4GUYCGMwAxnAL5wBmUEIcA5AAIDOwAdiQDYHBQD0MdBgBGEAB4MAUFP784AOiVwAQgRYYZJCOxbw1GgMJ1I7DOxi4SACwwgCcALwphCgPIiAVhhIwAFAhScMFwQpgAXC6YCvhQHADmfgCUWEEhwAAmkWjRsQnJqSFw7AQgGNmueeyJKVLkSTJyisrG4DrmMI3yhlAYRBhw2uxUwPGhwnAA7tak1nAArhoscPHmGHFkBFDx82UWy1TQcH02cGC0mLCoClIcMOtUBCQDPX33raYdVrb2xsOjAB4cgNnHo4tUcABJDJOOBghIAPkQaWCmUi0JR4wiUWwdSkQz0cFe-Q+7Qs3zsDmcQImGHE93YGWW8IhcGhx3p5iZcJg4PiCL8QxG8UixPeJjJlhslL+wppmChGQRSScSMCRV6MHmUHYOIUpLY9z8AG1MUUDBhSWZydL7IUiiFge4vD5-OqHR60XqADLAe5QAhcQU6YUKTK1D0esLlH1+9aB4P-eIKaMRyP1e3BAC6SQA3HV87J5EoFHAAKosAirLQ6QnIWFiy0S61Sn4EAKYr0Me56BiZrExhjqDLDvt1BpSaMoWH4IikAHRiBUFAImRSIA)

```typescript
import { Type, Static } from '@sinclair/typebox'

// ... Base

const BaseComponentSchema = Type.Object({
    type: Type.String(),
    id: Type.String(),
    name: Type.String(),
})

// ... Component

// Create config type which uses generic arguments for each property.
interface CreateComponentSchemaConfig<Type = string, Id = string> {
    id: Id
    type: Type,
}
const CreateComponentSchema = <Type extends string, Id extends string>(config: CreateComponentSchemaConfig<Type, Id>) => {
    return Type.Composite([
        BaseComponentSchema,
        Type.Object({
            id: Type.Literal(config.id),
            type: Type.Literal(config.type),
        }),
    ]);
};

// ... Usage

const T = CreateComponentSchema({
    id: 'test',
    type: 'asdasd',
})

type T = Static<typeof T>
```
The trick is encoding the string literals as generic arguments for the `config` interface, this enables TS to propagate the types into the Literal values.

Hope this helps
S

**Reply by alexey13** [reply]
Thank you!
